File: android/src/main/kotlin/io/sentient/android/settings/SessionLabel.kt
// ---------------------------------------------------------------------------
// SessionLabel — human-readable label for a vitals session row.
//
// The newest session is "This session"; older ones get a relative day prefix
// ("Today" / "Yesterday" / a short date) plus a wall-clock time ("9:43 PM").
// Pure over epoch-ms + the session info; formatting uses java.util.Calendar /
// SimpleDateFormat (android UI code, not commonMain — purity rule N/A here).
// ---------------------------------------------------------------------------
package io.sentient.android.settings

import io.sentient.mobilesdk.vitals.VitalsSessionInfo
import java.text.SimpleDateFormat
import java.util.Calendar
import java.util.Date
import java.util.Locale

private const val DAY_MS = 86_400_000L
private const val THIS_SESSION = "This session"
private const val UNKNOWN = "Unknown time"

/**
 * Label for one session row. The newest ([isNewest]) renders "This session"; others
 * render "<Today|Yesterday|MMM d> <h:mm a>", e.g. "Yesterday 2:07 PM".
 */
fun sessionLabel(info: VitalsSessionInfo, nowMs: Long, isNewest: Boolean): String {
    if (isNewest) return THIS_SESSION
    val ms = info.sessionStartMs
    if (ms <= 0L) return UNKNOWN
    return "${dayPrefix(nowMs, ms)} ${timeFormat().format(Date(ms))}"
}

private fun dayPrefix(nowMs: Long, ms: Long): String {
    val today = dayIndex(nowMs)
    return when (dayIndex(ms)) {
        today -> "Today"
        today - 1 -> "Yesterday"
        else -> dateFormat().format(Date(ms))
    }
}

// Local-calendar day index so "Today"/"Yesterday" respect the device timezone
// (a raw ms/DAY_MS bucket would be UTC-based and wrong near midnight).
private fun dayIndex(ms: Long): Long {
    val cal = Calendar.getInstance().apply { timeInMillis = ms }
    val offsetMs = (cal.get(Calendar.ZONE_OFFSET) + cal.get(Calendar.DST_OFFSET)).toLong()
    return (ms + offsetMs) / DAY_MS
}

private fun timeFormat() = SimpleDateFormat("h:mm a", Locale.getDefault())

private fun dateFormat() = SimpleDateFormat("MMM d", Locale.getDefault())

File: ios/App/Settings/Diagnostics/SessionLabel.swift
// ---------------------------------------------------------------------------
// SessionLabel — human-readable label for a vitals session row. Swift mirror of
// Android's SessionLabel (settings/SessionLabel.kt).
//
// The newest session is "This session"; older ones get a relative day prefix
// ("Today" / "Yesterday" / a short date) plus a wall-clock time ("9:43 PM").
// Pure over epoch-ms + the session info; uses Calendar / DateFormatter in the
// device timezone so "Today" / "Yesterday" are correct near midnight.
// ---------------------------------------------------------------------------
import Foundation

enum SessionLabel {
    private static let thisSession = "This session"
    private static let unknown = "Unknown time"

    /// Label for one session row. The newest (`isNewest`) renders "This session";
    /// others render "<Today|Yesterday|MMM d> <h:mm a>", e.g. "Yesterday 2:07 PM".
    static func label(sessionStartMs: Int64, nowMs: Int64, isNewest: Bool) -> String {
        if isNewest { return thisSession }
        if sessionStartMs <= 0 { return unknown }
        let date = Date(timeIntervalSince1970: Double(sessionStartMs) / 1000)
        return "\(dayPrefix(nowMs: nowMs, ms: sessionStartMs)) \(timeFormatter.string(from: date))"
    }

    private static func dayPrefix(nowMs: Int64, ms: Int64) -> String {
        let cal = Calendar.current
        let now = Date(timeIntervalSince1970: Double(nowMs) / 1000)
        let then = Date(timeIntervalSince1970: Double(ms) / 1000)
        if cal.isDate(then, inSameDayAs: now) { return "Today" }
        if let yesterday = cal.date(byAdding: .day, value: -1, to: now),
           cal.isDate(then, inSameDayAs: yesterday) {
            return "Yesterday"
        }
        return dateFormatter.string(from: then)
    }

    private static let timeFormatter: DateFormatter = {
        let f = DateFormatter()
        f.setLocalizedDateFormatFromTemplate("jm")
        return f
    }()

    private static let dateFormatter: DateFormatter = {
        let f = DateFormatter()
        f.setLocalizedDateFormatFromTemplate("MMMd")
        return f
    }()
}
